crates/synthesis/src/qsd.rs

```diff
@@ -0,0 +1,800 @@
+    (a1, a2, b, c)
+}
+
+/// Verify ZXZ decomposition gives the same unitary
+#[allow(clippy::toplevel_ref_arg)]
+fn _zxz_decomp_verify(
+    mat: &DMatrix<Complex64>,
+    a1: &DMatrix<Complex64>,
+    a2: &DMatrix<Complex64>,
+    b: &DMatrix<Complex64>,
+    c: &DMatrix<Complex64>,
+) -> bool {
+    let n = mat.shape().0 / 2;
+    let iden = DMatrix::<Complex64>::identity(n, n);
+
+    let a_block = stack![a1, 0; 0, a2];
+
+    let b1 = iden.clone() + b;
+    let b2 = iden.clone() - b;
+    let b_block = stack![b1, b2; b2, b1];
+
+    let c_block = stack![iden, 0; 0, c];
+
+    let mat_check = a_block * b_block * c_block * Complex64::from(0.5);
+
+    const EPS: f64 = 1e-7;
+    abs_diff_eq!(mat, &mat_check, epsilon = EPS)
+}
+
+///  Decompose a generic multiplexer.
+///
+///        ────□────
+///         ┌──┴──┐
+///       /─┤     ├─
+///         └─────┘
+///
+/// represented by the block diagonal matrix
+///
+///         ┏         ┓
+///         ┃ um0     ┃
+///         ┃     um1 ┃
+///         ┗         ┛
+///
+/// to
+///            ┌───┐
+///     ───────┤ Rz├──────
+///       ┌───┐└─┬─┘┌───┐
+///     /─┤ w ├──□──┤ v ├─
+///       └───┘     └───┘
+///
+/// where v and w are general unitaries determined from decomposition.
+///
+/// .. note::
+///
+///     When the original unitary is controlled then the default value of ``opt_a2 = False``
+///     as we start with the demultiplexing step that does not work with the optimization A.2 of [1, 2].
+fn demultiplex(
+    um0: &DMatrix<Complex64>,
+    um1: &DMatrix<Complex64>,
+    opt_a1: bool,
+    opt_a2: bool,
+    depth: usize,
+    _ctrl_index: Option<usize>,
+    vw_type: VWType,
+) -> PyResult<(CircuitData, DMatrix<Complex64>, DMatrix<Complex64>)> {
+    let um0 = closest_unitary(um0.clone());
+    let um1 = closest_unitary(um1.clone());
+
+    let dim = um0.shape().0 + um1.shape().0;
+    let num_qubits = dim.ilog2() as usize;
+    let _ctrl_index = _ctrl_index.unwrap_or_else(|| num_qubits - 1);
+    let layout: Vec<Qubit> = (0.._ctrl_index)
+        .chain(_ctrl_index + 1..num_qubits)
+        .chain([_ctrl_index])
+        .map(Qubit::new)
+        .collect();
+    let um0um1 = um0.clone() * um1.adjoint();
+    let (eigvals, vmat) = if is_hermitian_matrix(&um0um1) {
+        let eigh = um0um1.symmetric_eigen();
+        let evals = eigh.eigenvalues;
+        let eigvals = evals.map(|x| Complex64::new(x, 0.));
+        let orthonormal_eigenvectors = QR::new(eigh.eigenvectors).q();
+        (eigvals, orthonormal_eigenvectors)
+    } else {
+        let schur = nalgebra::linalg::Schur::try_new(um0um1, 1e-12, 100000).unwrap();
+        let (vmat, evals) = schur.unpack();
+        let eigvals = evals.diagonal();
+        (eigvals, vmat)
+    };
+    let d_values: DVector<Complex64> = eigvals.map(|x| x.sqrt());
+    let d_mat: DMatrix<Complex64> = DMatrix::from_diagonal(&d_values);
+    let wmat = d_mat.clone() * vmat.adjoint() * um1.clone();
```

Comment:
You can avoid cloning here
```suggestion
    let wmat = &d_mat * vmat.adjoint() * um1;
```